crates/codegen/language/definition/src/compiler/versions.rs

```diff
@@ -0,0 +1,215 @@
+use semver::Version;
+use std::{
+    cmp::{max, min},
+    fmt::Display,
+    iter::once,
+};
+
+const MAX_VERSION: Version = Version::new(u64::MAX, u64::MAX, u64::MAX);
+
+#[derive(Clone, Debug)]
+pub struct VersionRange {
+    pub inclusive_start: Version,
+    pub exclusive_end: Version,
+}
+
+impl VersionRange {
+    pub fn starting_from(inclusive_start: &Version) -> Self {
+        return Self {
+            inclusive_start: inclusive_start.to_owned(),
+            exclusive_end: MAX_VERSION,
+        };
+    }
+
+    pub fn between(inclusive_start: &Version, exclusive_end: &Version) -> Self {
+        if inclusive_start <= exclusive_end {
+            return Self {
+                inclusive_start: inclusive_start.to_owned(),
+                exclusive_end: exclusive_end.to_owned(),
+            };
+        } else {
+            // This is an invalid state, which we still produce errors for.
+            // However, for validation to continue, let's correct the order:
+            return Self {
+                inclusive_start: exclusive_end.to_owned(),
+                exclusive_end: inclusive_start.to_owned(),
+            };
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        return self.inclusive_start == self.exclusive_end;
+    }
+}
+
+impl Display for VersionRange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.inclusive_start == MAX_VERSION {
+            write!(f, "MAX")?;
+        } else {
+            write!(f, "{}", self.inclusive_start)?;
+        }
+
+        write!(f, "..")?;
+
+        if self.exclusive_end == MAX_VERSION {
+            write!(f, "MAX")?;
+        } else {
+            write!(f, "{}", self.exclusive_end)?;
+        }
+
+        return Ok(());
+    }
+}
+
+#[derive(Debug)]
+pub struct VersionSet {
+    ranges: Vec<VersionRange>,
+}
+
+impl VersionSet {
+    pub fn empty() -> Self {
+        return Self { ranges: vec![] };
+    }
+
+    pub fn single(range: VersionRange) -> Self {
+        let mut instance = Self::empty();
+        instance.add(&range);
+        return instance;
+    }
+
+    pub fn union(ranges: impl IntoIterator<Item = VersionRange>) -> Self {
+        let mut instance = Self::empty();
+
+        for range in ranges.into_iter() {
+            instance.add(&range);
+        }
+
+        return instance;
+    }
+
+    pub fn is_empty(&self) -> bool {
+        return self.ranges.is_empty();
+    }
+
+    pub fn add(&mut self, range: &VersionRange) {
+        if range.is_empty() {
+            return;
+        }
+
+        let mut first_iter = self.ranges.iter().peekable();
+        let mut second_iter = once(range).into_iter().peekable();
+        let mut ranges = vec![];
+
+        loop {
```

Comment:
It took me a while to understand how we're managing the two streams and conditionally consume the elements from either or both. 

I'm not sure what the exact semantics should be when it comes to merging disjoint ranges or normalizing the representation. Since this is "pure" and very mechanical code, it'd be great to have some unit tests to serve as an example.

For instance, I expected the following result to be an empty set but it seems not to be:
```rust
let mut example = VersionSet::empty();
example.add(&range!(0,8,0 till 0,9,0));
example.add(&range!(1,0,0 till 1,1,0));

example.add(&range!(0,9,0 till 1,0,0));
dbg!(&example.difference(&VersionSet::single(range!(0,8,0 till 1,1,0))));
// &example.difference(&VersionSet::single(range!(0, 8, 0 till 1, 1, 0))) = VersionSet {
//     ranges: [
//         VersionRange {
//             inclusive_start: Version {
//                 major: 1,
//                 minor: 0,
//                 patch: 0,
//             },
//             exclusive_end: Version {
//                 major: 1,
//                 minor: 1,
//                 patch: 0,
//             },
//         },
//     ],
// }
```
Maybe simplifying the zipping/state management would help catch that?

Reply:
So is this expected? Do you want to leave that for a follow-up?

Reply:
Good catch. Fixed the corner case and added tests. Bug existed since the legacy implementation in `crates/codegen/schema/src/validation/visitors/version_set.rs`, so I'm fixing that as well.

Luckily, it didn't miss anything, and the grammar is still correct.

Reply:
Good catch. Fixed the corner case and added tests.

Luckily, it didn't miss anything, and the grammar is still correct.